xarray/backends/common.py

```diff
@@ -503,3 +503,43 @@ def assert_open(self):
         if not self._isopen:
             raise AssertionError('internal failure: file must be open '
                                  'if `autoclose=True` is used.')
+
+
+class PickleByReconstructionWrapper(object):
+
+    def __init__(self, opener, *args, **kwargs):
+
+        self.opener = opener
+        self.open_args = args
+        self.open_kwargs = kwargs
+
+        self._ds = None
+        self._isopen = False
+
+    @property
+    def value(self):
+        if self._ds is not None and self._isopen:
+            return self._ds
+        self._ds = self.opener(*self.open_args, **self.open_kwargs)
+        self._isopen = True
+        return self._ds
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        del state['_ds']
+        del state['_isopen']
+        if self.open_kwargs.get('mode', None) == 'w':
+            # file has already been created, don't override when restoring
+            state['open_kwargs']['mode'] = 'a'
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        self._ds = None
+        self._isopen = False
+
+    def __getitem__(self, key):
+        return self.value[key]
+
+    def __getattr__(self, name):
```

Comment:
This magic is convenient but I think it could possibly be a source of bugs. I would stick to requiring pulling out `.value`.